subsys/usb/usb_device.c

```diff
@@ -1377,8 +1377,10 @@ static void forward_status_cb(enum usb_dc_status_code status, const u8_t *param)
 	size_t size = (__usb_data_end - __usb_data_start);
 
 	for (size_t i = 0; i < size; i++) {
-		if (__usb_data_start[i].cb_usb_status) {
-			__usb_data_start[i].cb_usb_status(status, param);
+		struct usb_cfg_data *cfg = &__usb_data_start[i];
+
+		if (cfg->cb_usb_status_composite) {
+			cfg->cb_usb_status_composite(cfg, status, param);
```

Comment:
From my understanding, it need calls cfg->cb_usb_status too as follow if cb_usb_status_composite is NULL and cb_usb_status is not NULL.
		if (cfg->cb_usb_status_composite) {
			cfg->cb_usb_status_composite(cfg, status, param);
		} else if (cfg->cb_usb_status) {
			cfg->cb_usb_status(status, param);
		}
For example: mass1 + cdc_acm1 + cdc_acm2, the mass storage class driver use cb_usb_status and cdc class driver use cb_usb_status_composite.
Or we can remove cb_usb_status and replace cb_usb_status with cb_usb_status_composite, and rename cb_usb_status_composite as another name.

Please let me know if I am wrong.

Reply:
For composite configuration we need to add `cb_usb_status_composite` to MSC. Even if now only one instance of MSC can be enabled.

Probably renaming callback would be the right approach. So we might have common callback with `NULL` parameter for non composite configuration in place of `cfg`.

Reply:
Yes, it is better to rename the callback.
Do you plan to add device instance (struct device *dev ) to all USB class drivers? for example: MSC.
If you plan add it, is it better to add one (struct device *dev) parameter to the follow callback functions?
usb_request_handler class_handler;
usb_request_handler vendor_handler;
usb_request_handler custom_handler;
usb_dc_status_callback cb_usb_status;
void (*cb_usb_status_composite)(struct usb_cfg_data *cfg,
					enum usb_dc_status_code cb_status,
					const u8_t *param);
And add the device_name (const char *dev_name) to struct usb_cfg_data.
After these changes, the usb_get_dev_data_by_cfg, usb_get_dev_data_by_iface and usb_get_dev_data_by_ep are not needed any more.

Reply:
@MarkWangChinese fc34152a3bc7cfa7e668d2bd4524d9727725fe99 renames callbacks. There is other PR  #13310. You can comment there, since this PR is merged already.